### ISSUE TITLE
Improve rate limiting

### ISSUE THREAD

**Issue by yonjah** [nodemailer/wildduck#68]
I was going over the security wiki https://github.com/nodemailer/wildduck/wiki/Security-implementation
and saw that rate limiting is done per user account.
Reading the code it seem like limiting is only happening if the user exists.

I can think of two issues that this might cause -
User enumeration - you can send a login request multiple time and if you hit the rate limiting you know the user actually exist.
Denial of service - you can keep sending fake login attempts to the server preventing the user from logging in.

I think this can be improve by first limiting auth by ip address as well so a single ip can only do x login attempts. if x is much smaller that he rate limit on the user account it will make both attacks a harder requiring sending requests from multiple ips to reach the rate limit of a single account.

**Comment by andris9** [comment]
I changed rate limiting in 97af5287e49a083186bfc75faa1576c0439a0361 as such:

  * non existent users are also rate limited, so it would not be possible to enumerate users
  * IP address is used as part of the rate limit key, so rate limiting applies to the same username:ip pair, not just to a username

I have mostly seen two kind of issues with rate limiting:

  1. Denial of service, where entire account blocks for a long time after x-failed attempts. Someone intentionally used invalid password to lock an user out of his email account. This is the reason why I decreased the lock time window, to a small length (60 seconds counting from first invalid attempt) as it would prevent brute force attacks while not locking out the user for too long time. It does not prevent malicious denial of service though, so now I added the IP check as well
  2. Slow password crawling from a large array of IP addresses doing limited authentication attempts over a long period like half a year. This is not mitigated from current rate limiting system at all (invalid authentication attempts are quite rare)
